src/test/java/com/google/cloud/tools/managedcloudsdk/internal/download/StandardDownloaderTest.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.download;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.nio.charset.Charset;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.Path;
+import org.apache.commons.io.FileUtils;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+public class StandardDownloaderTest {
+
+  @Rule public TemporaryFolder tmp = new TemporaryFolder();
+
+  @Test
+  public void testValidateAndCreateFiles_targetDirectoryIsFile()
+      throws IOException, InterruptedException {
+    Path fileWithBadParent = tmp.newFile().toPath().resolve("xyz");
+
+    StandardDownloader downloader = new StandardDownloader(null, fileWithBadParent, null, null);
+    try {
+      downloader.call();
+      Assert.fail("NotDirectoryException expected but not thrown");
+    } catch (NotDirectoryException nde) {
+      // pass
+    }
+  }
+
+  @Test
+  public void testValidateAndCreateFiles_fileAlreadyExists()
+      throws IOException, InterruptedException {
+    Path fileThatExists = tmp.newFile().toPath();
+
+    StandardDownloader downloader = new StandardDownloader(null, fileThatExists, null, null);
+    try {
+      downloader.call();
+      Assert.fail("FileAlreadyExists expected but not thrown");
+    } catch (FileAlreadyExistsException faee) {
+      // pass
+    }
+  }
+
+  @Test
+  public void testDownloadURL_createsNewDirectory() throws IOException, InterruptedException {
+    Path destination = tmp.getRoot().toPath().resolve("dir-to-create").resolve("destination-file");
+    Path testSourceFile = createTestRemoteResource(1);
+    URL fakeRemoteResource = testSourceFile.toUri().toURL();
+
+    StandardDownloader downloader =
+        new StandardDownloader(fakeRemoteResource, destination, null, null);
+
+    downloader.call();
```

Comment:
I think my original comment got outdated.

Wrap with a check for the return?

`Assert.assertEquals(destination, downloader.call());`

Ditto below.